src/rust/engine/graph/src/lib.rs

```diff
@@ -75,77 +149,374 @@ pub struct Entry<N: Node> {
+  /// Spawn the execution of the node on an Executor, which will cause it to execute outside of
+  /// the Graph lock and call back into the graph lock to set the final value.
   ///
-  fn state<C>(&mut self, context: &C, entry_id: EntryId) -> EntryStateField<N::Item, N::Error>
+  fn run<C>(
+    context: &C,
+    entry_key: &EntryKey<N>,
+    entry_id: EntryId,
+    run_token: RunToken,
+    generation: Generation,
+    previous_dep_generations: Option<Vec<Generation>>,
+    previous_result: Option<Result<N::Item, N::Error>>,
+  ) -> EntryState<N>
   where
     C: NodeContext<Node = N>,
   {
-    if let Some(ref state) = self.state {
-      state.field.clone()
-    } else {
-      let start_time = Instant::now();
-      let state = match &self.node {
-        &EntryKey::Valid(ref n) => {
-          // Wrap the launch in future::lazy to defer it until after we're outside the Graph lock.
-          let context = context.clone_for(entry_id);
-          let node = n.clone();
-          future::lazy(move || node.run(context)).to_boxed()
+    // Increment the RunToken to uniquely identify this work.
+    let run_token = run_token.next();
+    match entry_key {
+      &EntryKey::Valid(ref n) => {
+        let context2 = context.clone_for(entry_id);
+        let node = n.clone();
+
+        context.spawn(future::lazy(move || {
+          // If we have previous result generations, compare them to all current dependency
+          // generations (which, if they are dirty, will cause recursive cleaning). If they
+          // match, we can consider the previous result value to be clean for reuse.
+          let was_clean = if let Some(previous_dep_generations) = previous_dep_generations {
+            context2
+              .graph()
+              .dep_generations(entry_id, &context2)
+              .then(move |generation_res| match generation_res {
+                Ok(ref dep_generations) if dep_generations == &previous_dep_generations => {
+                  // Dependencies have not changed: Node is clean.
+                  Ok(true)
+                }
+                _ => {
+                  // If dependency generations mismatched or failed to fetch, re-run the Node.
+                  Ok(false)
+                }
+              })
+              .to_boxed()
+          } else {
+            future::ok(false).to_boxed()
+          };
+
+          // If the Node was clean, complete it. Otherwise, re-run.
+          was_clean.and_then(move |was_clean| {
+            if was_clean {
+              // No dependencies have changed: we can complete the Node without changing its
+              // previous_result or generation.
+              context2
+                .graph()
+                .complete(&context2, entry_id, run_token, None);
+              future::ok(()).to_boxed()
+            } else {
+              // The Node needs to (re-)run!
+              let context = context2.clone();
```

Comment:
Definitely don't re-use the name `context` here, as it's actually different from `context` in the enclosing scope

Reply:
Fixed.